addon/services/scroll-activity.js

```diff
@@ -1,63 +1,96 @@
 import $ from 'jquery';
 import Evented from 'ember-evented';
 import Service from 'ember-service';
-import { A } from 'ember-array/utils';
-import { bind } from 'ember-runloop';
-import { isPresent } from 'ember-utils';
+import run from 'ember-runloop';
 
-export default Service.extend(Evented, {
-  _animationFrame: null,
+/*
+ * Polling uses rAF and/or a setTimeout at 16ms, however rAF will run in the
+ * microtask queue and might fire just after Ember's render step occurred.
+ * By enforcing that the interval between a poll and the previous must be
+ * below a reasonable number, we can be reasonably sure the main UI
+ * thread didn't just do a lot of work.
+ *
+ * This number show be above the minimum polling period (16ms)
+ */
+const MAX_POLL_PERIOD = 32;
 
-  subscribers: null,
+export default Service.extend(Evented, {
 
   init() {
     this._super(...arguments);
 
-    this.set('subscribers', A());
-    this.subscribe(document, document);
+    this._animationFrame = null;
+    this._subscribers = [];
+    this._lastCheckAt = new Date();
+    this.subscribe(document, document, () => {}, false);
+
     this._pollScroll();
   },
 
-  subscribe(target, element, callback = function () {}) {
+  subscribe(
+    target,
+    element,
+    callback=() => {},
+    highPriority=true
+  ) {
     if (!element.scrollTop) { // a DOM element instead of a jQuery object
       element = $(element);
     }
-    this.get('subscribers').pushObject({
+    this._subscribers.push({
       target,
       element,
       callback,
-      scrollTop: element.scrollTop() // get scroll pos
+      highPriority,
+      scrollTop: null
     });
   },
 
   unsubscribe(target) {
-    this.get('subscribers').removeObject(this.get('subscribers').findBy('target', target));
+    let { _subscribers: subscribers } = this;
+    for (let i=0;i<subscribers.length;i++) {
```

Comment:
Would prefer we be consistent in how we loop this object (unless there is a reason not to be)...on line 75 it's a forEach, but here you use a basic for-loop.